### PR TITLE
WT-13533 Separate eviction from cache

### PR DESCRIPTION
**Pull request by Siddhartha8899**
---
As we progress with eviction modularising, it's been found that separating eviction into it's own struct is crucial for defining it's clear boundary.  

Currently, cache metadata, cache pool and eviction are usually defined within a single struct. However, based on the boundaries defined in the [1.1 doc](https://docs.google.com/document/d/1tCE9WzuKzwZKKCpKDj2bIS4YuiZNKIICOUw_utzUfZk/edit#heading=h.nhzw09t11ip). It is evident that eviction, in terms of its functionality, should be distinct from cache metadata and cache pool. Therefore, the is an effort to separate eviction-related fields into its own struct.

---

### CODE REVIEW

**Comment by Siddhartha8899** [comment]
Closing for now, in favour of incremental changes.

File: src/include/connection.h
Comment (ershov) [inline src/include/connection.h:504]: I wonder if you considered including the evict structure right in the parent rather than making it a separate allocation? What would be the trade-offs? Some pros and cons?

File: src/include/wt_internal.h
Comment (ershov) [inline src/include/wt_internal.h:521]: Can we add the `src` directory in the standard include search path and use `#include <evict/evict.h>`?

Comment (Siddhartha8899) [reply src/include/wt_internal.h:521]: Yes, that's a good idea. Given that we are moving away for `include/` it makes sense. I will leave it up to @ajmorton as he is planning on moving the header file from include/ to evict/ and I will be moving it back soon.

File: src/include/connection.h
Comment (Siddhartha8899) [reply src/include/connection.h:504]: evict struct in the parent to cache?

File: src/include/wt_internal.h
Comment (ajmorton) [reply src/include/wt_internal.h]: SGTM 👍 I'll take this up in the prototypes.py PR

File: dist/stat_data.py
Comment (ajmorton) [inline dist/stat_data.py:103]: @raviprakashgiri29 we've changed the name of this stat group but there's no change to the list of stats or their text descriptions. 
Does this impact your current T2 work and the `WiredTiger Eviction Cache State` category?

File: src/evict/evict_stat.c
Comment (ajmorton) [inline src/evict/evict_stat.c:120]: Question: Why did we add the word `cache` into this function name? We just split it out from the cache logic

File: src/include/btmem.h
Comment (ajmorton) [inline src/include/btmem.h:32]:
```suggestion
 * Helper: in order to read a Btree without triggering eviction we have to ignore the cache size and
```

File: src/include/cache.h
Comment (ajmorton) [inline src/include/cache.h:86]: Question: Are we tracking that we want to remove cache_pool in either a Jira ticket or a comment on the PM-3441 Tech design?

File: src/include/evict.h
Comment (ajmorton) [inline src/include/evict.h:148]: nit: Do we need `_CACHE_` in these macros? `WT_EVICT_CLEAN`/`WT_EVICT_DIRTY_HARD`/`WT_EVICT_URGENT` (for a few examples) seem readable to me.

File: src/evict/evict_conn.c
Comment (ajmorton) [inline src/evict/evict_conn.c:248]: Thanks for remembering this!

File: src/btree/bt_page.c
Comment (ajmorton) [inline src/btree/bt_page.c:118]:
```suggestion
    (void)__wt_atomic_add64(&S2C(conn)->cache->pages_inmem, 1);
```
and we can drop the local `cache` variable

File: src/evict/evict_lru.c
Comment (ajmorton) [inline src/evict/evict_lru.c:280]: nit:
```suggestion
    __wt_cond_signal(session, S2C(conn)->evict->evict_cond);
```
and drop the local `evict` variable

File: src/evict/evict_inline.h
Comment (ajmorton) [inline src/evict/evict_inline.h:278]: nit:
```suggestion
    return (__wt_cache_bytes_plus_overhead(cache, __wt_atomic_load64(&S2C(conn)->bytes_hs_dirty)) >=
```
and we don't need the local `cache` variable

File: dist/stat_data.py
Comment (ajmorton) [inline dist/stat_data.py:841]: This might be the change we discussed last week, but up on line 260 we have the `Cache and eviction statistics` stats with lines like `    CacheStat('cache_eviction_app_fail', 'page evict failures by application threads'),` which should now be EvictStats

**Review by ajmorton — APPROVED** [review]
LGTM! When you're ready add those last renaming changes and re-request my review.

File: dist/stat_data.py
Comment (raviprakashgiri29) [reply dist/stat_data.py:103]: @ajmorton Thanks for the heads up. This should not have any impact on t2 since the 'cache_walk' prefix and stat text descriptions remain unchanged.

File: src/evict/evict_stat.c
Comment (Siddhartha8899) [reply src/evict/evict_stat.c:120]: This is not the stat function that we split from cache_conn.c into evict_conn.c; that one is still neatly separated in `__wt_evict_stats_update`.

What we have here is the dhandle-level stat walk function, which updates both evict and cache stats together (EvictCacheWalkStats). I’ve changed the function name to better reflect this. We can consider separating it out in the future, but this logic hasn't seen much use in the past 10 years, so I thought to leave it as is (as it's much more complicated to seperate out neatly).

File: src/include/cache.h
Comment (Siddhartha8899) [reply src/include/cache.h:86]: No, we aren't currently tracking the removal of the cache pool. I can create a ticket to investigate this further. From what I’ve heard, it seems we plan to remove it unless disaggregated storage requires it.

Comment (Siddhartha8899) [reply src/include/cache.h:86]: https://jira.mongodb.org/browse/WT-13595 done!

File: src/include/evict.h
Comment (Siddhartha8899) [reply src/include/evict.h:148]: I’m open to removing it and can make the changes if you’d like. CACHE helps convey that we are evaluating things from the cache, but that’s somewhat implied anyway.

Let me know if you'd like me to remove it.

File: src/btree/bt_page.c
Comment (Siddhartha8899) [reply src/btree/bt_page.c:115]: Updated in [add50ca](https://github.com/wiredtiger/wiredtiger/pull/11037/commits/add50ca3177b3a8d182c20b2e499f52f2ee3b3d3)

File: src/evict/evict_lru.c
Comment (Siddhartha8899) [reply src/evict/evict_lru.c]: Updated in [a544057](https://github.com/wiredtiger/wiredtiger/pull/11037/commits/a5440573ac1f3c62af793041ae953a033d885ae3)

File: src/evict/evict_inline.h
Comment (Siddhartha8899) [reply src/evict/evict_inline.h:278]: but cache is used twice here?

File: src/evict/evict_stat.c
Comment (ajmorton) [reply src/evict/evict_stat.c:120]: SGTM 👍. Resolving

File: src/evict/evict_inline.h
Comment (ajmorton) [reply src/evict/evict_inline.h:278]: My bad, must've missed it 👀

File: src/include/evict.h
Comment (ajmorton) [reply src/include/evict.h:148]: Not enough to push for it. If you're happy with it let's keep it 👍 . Resolving

**Review by ajmorton — APPROVED** [review]
The new changes LGTM. @Siddhartha8899 can you confirm two things for me about the latest commits?
1. None of the stat descriptions have changed
2. The change to stat names is just "cache_eviction_*" to "eviction_*"

**Comment by Siddhartha8899** [comment]
> The new changes LGTM. @Siddhartha8899 can you confirm two things for me about the latest commits?
> 
> 1. None of the stat descriptions have changed
> 2. The change to stat names is just "cache_eviction__" to "eviction__"

Both the points are **True**

File: dist/stat_data.py
Comment (ershov) [inline dist/stat_data.py:288]: It's hard to track down changes here. Just want to make sure that changes to stats only touch internal names, not external representation or flags.

Comment (ershov) [reply dist/stat_data.py:288]: Ok I see it now from other comments.

Comment (Siddhartha8899) [reply dist/stat_data.py:288]: Yes, it touches only the internal names.

**Pull request merged**
